Add P2LB  module and feature split

cory-skeers opened the pull request:
Building block for an automated move from paragraphs to layout builder. Resolves #1803.

# To test
- `ddev blt ds --site brand.uiowa.edu`
- Observe there are no issues with site sync.
- `ddev drush @brand.local uli`
- Ensure that no P2LB functionality is present on the brand site.

<details>
<summary>Original post</summary>
Process is encapsulated in the sitenow_p2lb module, which adds a config menu at /admin/config/system/sitenow-p2lb. Currently select nodes (ids populated based on presence of paragraphs content on the page), and the option to delete all paragraphs content (tester, or to use following migrate) or update, which runs through the various functions within the main module to move content over. The moved content remains on the same page, so that internal links are preserved.

List of functionality considerations started on #1803.

Couple additional notes:
- Depending on resolution of #1555, some refactoring will be needed. Current build assumes all paragraphs on a site should be removed, which may not be the case.
</details>

cory-skeers committed:
Skeleton build of p2lb module.

cory-skeers committed:
Code cleanup.

cory-skeers committed:
Start building queries, loads, etc.

cory-skeers committed:
Build out methods a bit more.

cory-skeers committed:
Coding cleanup.

cory-skeers committed:
Coding cleanup.

cory-skeers committed:
Add paragraphs delete function. Start work on form function options for testing and/or UI use.

cory-skeers committed:
Forgot extra line at EoF...

cory-skeers committed:
Add paragraph delete function to UI.

cory-skeers committed:
Add paragraph update function to UI.

cory-skeers committed:
Start working on building block sections from paragraph col_widths.

cory-skeers committed:
Short array syntax. Code spacing.

cory-skeers committed:
Update text p2lb.

cory-skeers committed:
Function updates.

cory-skeers committed:
Remove whitespace

cory-skeers committed:
Add section column logic. Update section building.

cory-skeers committed:
Coding standards.

cory-skeers committed:
Coding standards.

cory-skeers committed:
Update determine_columns array handling.

cory-skeers committed:
Update todo...

cory-skeers committed:
Updates to array handling

cory-skeers committed:
Update to layout_builder__layout setting on node.

cory-skeers committed:
Condition checks. Work toward getting paragraphs within sections processed correctly.

cory-skeers committed:
Start building out card migrate process.

cory-skeers committed:
Quick fix for sections without admin titles.

cory-skeers committed:
Add checks for null values.

cory-skeers committed:
Cleaning up some method calls.

cory-skeers committed:
Add region handling for multi-col sections.

cory-skeers committed:
Reordering, cleaning up some code.

cory-skeers committed:
Documentation, code cleanup.

cory-skeers committed:
Documentation, code cleanup.

cory-skeers committed:
Documentation, code cleanup.

cory-skeers committed:
Documentation, code cleanup.

cory-skeers committed:
Debugging. Set col_width string, not array.

cory-skeers committed:
Refactor handling of card body content/subtitle.

cory-skeers committed:
Code cleanup / standards.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Minor form tweak

cory-skeers committed:
Refactor col width checks to only recurse; bug fixes.

cory-skeers committed:
Main merge.

cory-skeers committed:
Update to use StringTranslationTrait

cory-skeers committed:
Add return statement to avoid extra looping in addition to recurse in column setup.

cory-skeers committed:
Update card and text block label handling.

cory-skeers committed:
Initial work at passing section styles from p to lb.

pyrello commented on docroot/modules/custom/sitenow_p2lb/src/Form/P2LbSettingsForm.php:
```diff
@@ -0,0 +1,105 @@
+    }
+    // @todo Option to remove paragraphs after migrate, or review first?
+    return $form_state;
+  }
```
We might want to look at using the Queue API for processing a large amount of nodes, if that situation exists: https://www.sitepoint.com/drupal-8-queue-api-powerful-manual-and-cron-queueing/

cory-skeers committed:
Bug fixes in the style applicator.

cory-skeers committed:
Quick fix to array handler.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Load node rather than db query to fix old revision bug.

cory-skeers committed:
Updates to how para/node child ids are grabbed.

cory-skeers committed:
Remove extra whitespace.

cory-skeers committed:
Bug fix to paragraph iteration when creating lb sections.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Remove default lb section that holds the paragraphs content.

cory-skeers committed:
Section background image handling.

cory-skeers committed:
Coding standards.

cory-skeers committed:
Merge main updates.

cory-skeers committed:
Remove invalid switch cases

cory-skeers committed:
Blockquote handling

cory-skeers committed:
Whitespace fixes

cory-skeers committed:
Initial pass at getting webform blocks to work.

cory-skeers committed:
Webform handling bug fixes.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Start of carousel->slider handling.

cory-skeers committed:
Extract a headline and build a paragraph uiowa_slide entity for each carousel item.

cory-skeers committed:
Updates to individual slide handling, array handling.

cory-skeers committed:
Code cleanup.

cory-skeers committed:
Fix to child field attachment on slider.

cory-skeers committed:
Quickfix for carousel items without a header.

cory-skeers committed:
Change to multi-select for sites with higher pagecounts.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Fixed new PHPCS errors.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Minor update to add access dependency through usage for programmatically added blocks.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Add todos/docs. Update section classes.

cory-skeers committed:
Update block usage for banner blocks created out of section bg images.

cory-skeers committed:
Update section bg images migrate to use banners, not heroes.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Added @todo comment during card migration.

pyrello committed:
Added @todo comment during text area migration.

pyrello committed:
Merge branch 'main' into p2lb

richardbporter committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Add magic button. Some updates to get it closer to being up-to-date.

cory-skeers committed:
PHPCS. Unset useless form button.

cory-skeers committed:
Add default card styles as an example to how they could be added for all block types.

cory-skeers committed:
Fix card excerpt bug. Add style checker for text areas.

cory-skeers committed:
Build uiowa_collection from accordion and set view_mode.

cory-skeers committed:
Start building accordion children.

cory-skeers committed:
Finish up accordion child mappings.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

Commit:
Only process carousel image caption if not empty

cory-skeers committed:
Add basic confirmation prior to AMAZEMENT. And an opportunity to not be amazed.

Commit:
Set text to an empty array when banner text is null

This handles converting a section with a background image but no text.
Example: https://hawkeyemarchingband.uiowa.edu/node/11

Commit:
Fix typo

Commit:
phpcs

cory-skeers committed:
Add/update some comments. Minor code cleanups.

cory-skeers committed:
PHPCS

cory-skeers committed:
Add parent id to collection item children.

cory-skeers committed:
Added framework for people views list blocks.

cory-skeers committed:
Framework of articles transfer.

cory-skeers committed:
Built framework of events block. Reorganized switch cases for alphabetical order now that more of them have been built out. Add todos.

cory-skeers committed:
Build framework for image gallery transfer.

cory-skeers committed:
Update people sort and results_per_page.

cory-skeers committed:
Headline handling for text blocks.

cory-skeers committed:
Clearing up some todos. Cleanup set checks. Verified background style names still relevant.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Added the views_bulk_operations module and started working on an action.

pyrello committed:
Fixed conflicts and finished merging 'main' branch.

pyrello committed:
Updated sitenow_p2lb_node_p2lb function to work with passed in entity. Updated action. Added admin view to manage conversion process.

pyrello committed:
VBO is not actually needed to be able to run the bulk action, so removed it. Added system action config for conversion. Updated action plugin and view to reflect changes.

pyrello committed:
Fixed PHPCS issues.

pyrello committed:
Minor refactor to change processing.

pyrello committed:
Fixed PHPCS error.

pyrello committed:
Added field for tracking revision ID when conversion happened.

pyrello committed:
Re-built the converter UI to use revisions instead of content so we can show drafts. Added code to set a new draft and store the revision ID of the node before it was converted to the conversion revision ID field.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
PHPCS fix.

pyrello committed:
Refactoring to reduce some repetition.

pyrello committed:
Added check to ensure that value property of array is not be accessed if it doesn't exist. Fixed getting block bundle correctly. Fixed issue with variables not being set.

cory-skeers committed:
Merge branch 'main' into p2lb

pyrello committed:
Generate the header section when relevant. Added function comments.

cory-skeers committed:
Return a section object in place of array for the header section config

cory-skeers committed:
Make sure  is defined, even if NULL. @todo determine if we even need/want labels.

pyrello committed:
Shift strategy on pulling in content moderation and header sections to use already defined config.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Small refactor for adding default sections.

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Get featured content items as a field item list.

richardbporter committed:
Merge branch 'main' into p2lb

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Merged main branch, fixed conflicts.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Move P2LB features to a different feature split so it can be turned on independently of SiteNow V2.

pyrello committed:
Updated SiteNow V2 split definition.

pyrello committed:
Reverting some unintentional changes.

pyrello committed:
Update P2LB split to include a field that was excluded and make it stackable.

pyrello committed:
Merge branch 'main' into p2lb

pyrello committed:
Fixed PHPCS stuff.

pyrello committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Merge branch 'main' into p2lb

cory-skeers committed:
Move styles setting to separate helper function to ease updating.

cory-skeers committed:
Added clarifying comment.

joewhitsitt committed:
Merge branch 'main' into p2lb

joewhitsitt reviewed (approved):
It is not on, and I don't see any of it as a webmaster or admin
